Comorphisms/SuleCFOL2TPTP.hs

```diff
@@ -0,0 +1,682 @@
+{-# LANGUAGE MultiParamTypeClasses, TypeSynonymInstances, FlexibleInstances #-}
+{- |
+Module      :  ./Comorphisms/SuleCFOL2TPTP.hs
+Description :  Coding of a CASL subset into TPTP
+Copyright   :  (c) Eugen Kuksa and Till Mossakowksi
+License     :  GPLv2 or higher, see LICENSE.txt
+
+Maintainer  :  kuksa@iks.cs.ovgu.de
+Stability   :  provisional
+Portability :  non-portable (imports Logic.Logic)
+
+The translating comorphism from a CASL subset to TPTP.
+-}
+
+module Comorphisms.SuleCFOL2TPTP
+  ( suleCFOL2TPTP
+  ) where
+
+import Control.Exception (assert)
+import Control.Monad (foldM)
+
+import Logic.Logic as Logic
+import Logic.Comorphism
+
+import Common.AS_Annotation
+import Common.Id
+import Common.Result
+import Common.DocUtils
+import Common.ProofTree
+import Common.Utils
+import qualified Common.Lib.Rel as Rel
+import qualified Common.Lib.MapSet as MapSet
+import Control.Monad (liftM)
+
+import qualified Data.Map as Map
+import qualified Data.Set as Set
+
+import Data.List as List
+import Data.Maybe
+import Data.Char
+import Data.Function
+import Numeric (showHex)
+
+import CASL.Logic_CASL
+import CASL.AS_Basic_CASL as CAS
+import CASL.Cycle
+import CASL.Sublogic as SL
+import CASL.Sign as CSign
+import CASL.MapSentence
+import CASL.Morphism
+import CASL.Quantification
+import CASL.Overload
+import CASL.Utils
+import CASL.Inject
+import CASL.Simplify
+import CASL.ToDoc
+
+import TPTP.AS as TAS
+import TPTP.Morphism as TMorphism
+import TPTP.Sign as TSign
+import TPTP.Logic_TPTP
+import TPTP.Sublogic
+
+data TPTP_FOF = TPTP_FOF
+
+instance Show TPTP_FOF where
+  show TPTP_FOF = "TPTP_FOF"
+
+-- | The identity of the comorphisms
+data GenSuleCFOL2TPTP a = GenSuleCFOL2TPTP a deriving Show
+
+suleCFOL2TPTP :: GenSuleCFOL2TPTP TPTP_FOF
+suleCFOL2TPTP = GenSuleCFOL2TPTP TPTP_FOF
+
+-- | TPTP theories
+type TPTPTheory = (TSign.Sign, [Named Sentence])
+
+instance Show a => Language (GenSuleCFOL2TPTP a) where
+  language_name (GenSuleCFOL2TPTP a) = "CASL2" ++ show a
```

Comment:
I copied and adjusted https://github.com/spechub/Hets/blob/430ab7208b425d007aa2d7e035eb0a3329da20bc/Comorphisms/SuleCFOL2SoftFOL.hs#L152-L153 because I don't know the mechanics of it.

Reply:
the reason for genericity there is that there are two variants of the comorphism. Here, it seems that we have only one variant, so we can make things simpler (look at other comorphisms).

Reply:
This should indeed be simplified as suggested by Till.

Reply:
But other than my very minor comment below, the code looks good and can be merged. I also tested the translation on the test file provided.

Reply:
Done with 6b968b5ef.

Reply:
Well, but we only have `GenSuleCFOL2TPTP TPTP_FOF`, and we don't plan to have versions of this comorphism, that would justify using a variable `a`, that can be filled in later with other values. Why not have instead of 
```newtype GenSuleCFOL2TPTP a = GenSuleCFOL2TPTP a```, 
```data GenSuleCFOL2TPTP = GenSuleCFOL2TPTP```?

Reply:
yes, good idea

Reply:
You're right. Done with f3dbd0334.

Reply:
Looks good to me now. :+1: